### PR TITLE
PiP enter/leave programmatically

### PR DESCRIPTION
**Pull request by rolandkakonyi**
## Description

It is currently not possible to trigger entering or leaving PiP mode from a React Native application.

## Changes

1. Introduced `PlayerView.isPictureInPictureRequested` to trigger entering or leaving PiP
2. Extended PiP example screen to feature a button for testing this

## Manual testing
Open the `Basic Picture in Picture` screen and use the button on the top-right.

## Checklist
- [x] 🗒 `CHANGELOG` entry

### CODE REVIEW

File: example/src/App.tsx [hunk 1]
```diff
@@ -204,7 +205,11 @@ export default function App() {
         <RootStack.Screen
           name="BasicPictureInPicture"
           component={BasicPictureInPicture}
-          options={{ title: 'Basic Picture in Picture' }}
+          options={{
+            title: 'Basic Picture in Picture',
+            // eslint-disable-next-line react/no-unstable-nested-components
```
Comment (zigavehovec) [inline example/src/App.tsx:210]: This looked fishy to me, but after looking at the [component documentation](https://reactnavigation.org/docs/header-buttons/) I guess is this is the way to do it.

File: ios/RNPlayerViewManager.swift [hunk 2]
```diff
@@ -105,6 +105,28 @@ class RNPlayerViewManager: RCTViewManager {
...
+        bridge.uiManager.addUIBlock { [weak self] _, views in
+            guard
+                let self,
+                let view = views?[viewId] as? RNPlayerView
+            else {
+                return
+            }
+            guard let playerView = view.playerView else {
+                return
+            }
```
Comment (zigavehovec) [inline ios/RNPlayerViewManager.swift:118]: Is this not possible?

```suggestion
            guard
                let self,
                let view = views?[viewId] as? RNPlayerView,
                let playerView = view.playerView
            else {
                return
            }
```

**Review by zigavehovec — COMMENTED** [review]
Looks very goo, I added some comments, but no real blockers 👌

File: example/src/App.tsx (hunk 1 above)
Comment (rolandkakonyi) [reply example/src/App.tsx:210]: I know and I couldn't find any other way where the warning doesn't show up.

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,6 +8,7 @@
 - `PictureInPictureConfig` to `PlayerView` to allow configuring the Picture in Picture behavior
 - `PictureInPictureConfig.shouldEnterOnBackground` to start PiP automatically when the app transitions to background
 - `onPictureInPictureAvailabilityChanged` event is now emitted on iOS and tvOS in addition to Android
+- Support for programmatic Picture in Picture request (`PlayerView.isPictureInPictureRequested`)
```
[the remark this answers is not in the document]
Comment (rolandkakonyi) [reply CHANGELOG.md:11]: Sounds good! Thanks for the suggestion!

File: ios/RNPlayerViewManager.swift (hunk 2 above)
Comment (rolandkakonyi) [reply ios/RNPlayerViewManager.swift]: fixed in https://github.com/bitmovin/bitmovin-player-react-native/pull/289/commits/5e20a616a1b0b225b5268e21e1188017625593bc

**Review by zigavehovec — COMMENTED** [review]
I tested the Android sample and found 2 issues.

**"Enter PiP" button not working correctly**
Steps to reproduce: 
1.) Click on Enter PiP - the app goes into PiP mode correctly
2.) Stop the PiP mode by either opening the app again from launcher or clicking on the go back to fullscreen button on the PiP interface
3.) Clicking on the Enter PiP button does nothing. After clicking one more time the app goes into PiP mode again correctly.

**Wrong app state after stopping PiP**
Another bug I noticed is when stopping the player by dragging the PiP interface down (stops the PiP and stops the playback). 
When opening the app again the app is in a weird state without a toolbar and a black player: 

image

File: android/src/main/java/com/bitmovin/player/reactnative/RNPlayerViewManager.kt [hunk 4]
```diff
@@ -196,6 +202,20 @@ class RNPlayerViewManager(private val context: ReactApplicationContext) : Simple
...
+        Handler(Looper.getMainLooper()).post {
+            with(view.pictureInPictureHandler ?: return@post) {
+                if (isPictureInPicture) {
+                    enterPictureInPicture()
+                } else {
+                    exitPictureInPicture()
+                }
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (rolandkakonyi) [reply android/src/main/java/com/bitmovin/player/reactnative/RNPlayerViewManager.kt]: improved Kotlin code style in cf28d6d

**Comment by 123mpozzi** [comment]
I found 2 issues:
- `onPictureInPictureEntered` and `onPictureInPictureExited` are not triggered on Android
- On Android `onPictureInPictureEnter` Event is not triggered with the button `Enter PiP`

**Comment by rolandkakonyi** [comment]
> I tested the Android sample and found 2 issues.
> 
> **"Enter PiP" button not working correctly** Steps to reproduce: 1.) Click on Enter PiP - the app goes into PiP mode correctly 2.) Stop the PiP mode by either opening the app again from launcher or clicking on the go back to fullscreen button on the PiP interface 3.) Clicking on the Enter PiP button does nothing. After clicking one more time the app goes into PiP mode again correctly.
> 
> **Wrong app state after stopping PiP** Another bug I noticed is when stopping the player by dragging the PiP interface down (stops the PiP and stops the playback). When opening the app again the app is in a weird state without a toolbar and a black player:
> 
> image

@zigavehovec I fixed missing events problems in 2bfc0e85c5ada0d6c1f8d2d1b8ca3597c9c98c29

Could you test again?

**Comment by rolandkakonyi** [comment]
> I found 2 issues:
> 
> * `onPictureInPictureEntered` and `onPictureInPictureExited` are not triggered on Android
> * On Android `onPictureInPictureEnter` Event is not triggered with the button `Enter PiP`

@123mpozzi I fixed missing events in 2bfc0e85c5ada0d6c1f8d2d1b8ca3597c9c98c29, could you test again?

**Comment by 123mpozzi** [comment]
Now `onPictureInPictureEnter` gets triggered twice when clicking the Player UI, but the exit event is fine.
The programmatic button is also fine.

File: android/src/main/java/com/bitmovin/player/reactnative/RNPlayerViewManager.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (zigavehovec) [reply android/src/main/java/com/bitmovin/player/reactnative/RNPlayerViewManager.kt]: You should extract the playerView into a variable here (the same as pictureInPictureHandler): https://github.com/bitmovin/bitmovin-player-react-native/commit/cf28d6dae9fad765fd0ff231a0067da50dbeb566#diff-1fe085284c1be1f053b637a06423c61b4c400f12ba0a50a30209089d998dd4e3R194
That way there is no need to unwrap it anymore later 😊

**Comment by zigavehovec** [comment]
> Could you test again?

I can confirm what @123mpozzi mentioned [here](https://github.com/bitmovin/bitmovin-player-react-native/pull/289#issuecomment-1764429975).

File: android/src/main/java/com/bitmovin/player/reactnative/RNPlayerViewManager.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (rolandkakonyi) [reply android/src/main/java/com/bitmovin/player/reactnative/RNPlayerViewManager.kt]: Fixed in 2468c01

**Comment by rolandkakonyi** [comment]
@zigavehovec @123mpozzi could you test again?

**Comment by zigavehovec** [comment]
> @zigavehovec @123mpozzi could you test again?

It looks like the first issue is fixed now! 👍

**Review by zigavehovec — APPROVED** [review]

**Comment by 123mpozzi** [comment]
It looks fine to me!

**Review by 123mpozzi — APPROVED** [review]

**Pull request merged**